### ISSUE TITLE
Response code for POST method should be 201 not 200

### ISSUE THREAD

**Issue by jannyHou** [loopbackio/loopback-next#788]
Response code for POST method should be 201 not 200

An old issue from LoopBack3: https://github.com/strongloop/loopback/issues/360

We should fix it in LoopBack4 as well.

### Notes from the spike

- Users should be able to customize the status code from the response object returned by the controller method, otherwise we should default to 201 - this will be implemented by #436.

- There are two approaches we can choose as implementation, but based on the spike findings and the consensus, pursue option 1 from below:
  - option 1: return the status code and header from the result of controller method directly
  - ~option 2: read the status code from OpenAPI spec and write it into the response in `rest/src/writer.ts`~, see the spike PR https://github.com/strongloop/loopback-next/pull/1260
  - There is already some comparison and discussion for the two approaches in PR https://github.com/strongloop/loopback-next/pull/1260 to start with, chose one approach and implement it. 

### Acceptance criteria

- [ ] Modify controllers in the `example/*` projects so that `create*` methods (typically mapped to `POST` verb) return `201 Created` instead of `200.
  - Update the OpenAPI Spec description of operation result, replace 200 to 201
  - Update the controller method implementation to return status code 201 and set `Location` header to the URL where the newly created model instance can be found
- [ ] Update CRUD Controller template used by CLI similarly
- [ ] If https://github.com/strongloop/loopback-next/issues/2736 has been done, then update the extension accordingly too.

**Comment by Cooker-Monster** [comment]
Thanks to reactivate this important point !

**Comment by bajtos** [comment]
I think this has two parts:

 - We need to figure out how to allow controllers to specify response status code (and response headers, while we are at this)

 - The CRUD controller template (see https://github.com/strongloop/loopback-next/issues/727) and the CRUD controller base class (see https://github.com/strongloop/loopback-next/pull/740) should return 201 from `create` operation

**Comment by jannyHou** [comment]
hi team, we are upgrading from swagger2 to openapi3, PR is ready for it https://github.com/strongloop/loopback-next/pull/916
There is an interface change of the `responseObject` in openapi3, I think it's better to implement this story for v3 directly instead of implement it with v2 then take another effort to upgrade to v3, thanks!

**Comment by kjdelisle** [comment]
Does the alteration to the responseObject formatting change the logic required to handle the return code in some way? I would've expected that the sequence would simply handle the success of a POST with a 201 code in the same way for v2 and v3; that is, I'd expect a property with the same name (probably `statusCode`) at the top level.

**Comment by jannyHou** [comment]
@kjdelisle good point, honestly I am not sure, and due to the effort needed in v2-v3, see comment:https://github.com/strongloop/loopback-next/issues/753#issuecomment-362054454, the PR won't be landed in a few days, so no sense to block this story :)

**Comment by bajtos** [comment]
@kjdelisle and me decided this does not really belong to MVP/Walking Skeleton.

**Comment by Cooker-Monster** [comment]
So, can you give a status of this issue please ? can we espect a fix on 4.x ?
whatever the version of Open Api specification, have a correct code for POST still make sense ;-)
But I'm not at the right position to mesure the effort... so it's easy to ask !

**Comment by kjdelisle** [comment]
Marking the item as non-MVP doesn't mean that we won't be doing it for the 4.x release, it just means that it's not a part of our initial push to get the framework into a stable state.

Our goal with the idea of MVP is that we want the framework to be in a position where:
- Community members can begin learning how it works in earnest
- Community members can begin to contribute fixes, as well as new features or plugins

The MVP is *not* the final release! It's an interim state between the constant flux of alpha and the near-completeness of a beta. We've set this goal to help stop us from getting too far ahead of ourselves with respect to what we want to see in the final release vs. what _should_ be there at a minimum.

**Comment by Cooker-Monster** [comment]
Ok fine ! thx

**Comment by Cooker-Monster** [comment]
Hi, any news on this topic please ?

**Comment by jannyHou** [comment]
@Cooker-Monster checking with my team, will let you know the update then.

**Comment by dhmlau** [comment]
@jannyHou , you've mentioned that this is related to the openapi-v3 work that you're doing.  Could you please add more details?  Thanks.

**Comment by jannyHou** [comment]
@Cooker-Monster @dhmlau sorry for the late reply!

I think there are two things related to the status code:

1. Generate the OpenAPI `response` specification
  - It would happen in `openapi-v3`, see [generating operation spec in file `controller-spec.ts`](https://github.com/strongloop/loopback-next/blob/master/packages/openapi-v3/src/controller-spec.ts#L77), we can detect the operation type and generating `response` accordingly. E.g. If an operation is `post`, the successful status code for it is `201`.
  - `schema` can be generated based on the controller function's metadata. This is off topic here, but to generate a complete response spec, we need both `statuscode` and `schema`, and also the content type.
2. Write the http response according to the `result` from the controller method.
  - It can use the OpenAPI path spec as the reference to decide the status code, which means it reads the `201` as the successful status code for a post method, write it to http response.
  - This would happen in `rest` package, [`src/writer.ts`](https://github.com/strongloop/loopback-next/blob/71e5af10365b7fa6277cad8f6a216439610f19df/packages/rest/src/writer.ts) file.

While I appreciate some opinion from @bajtos in case my understanding of 2 is not correct.

**Comment by dhmlau** [comment]
Discussing with @raymondfeng yesterday, the goal here is that given the default response code being `200`, how we can let users to customize the response status code and even body.  One possibility is to see what we can do with the `@post` decorators.

**Comment by dhmlau** [comment]
I don't think this task has a clear acceptance criteria.  Moving it back to `needs discussion` column.

**Comment by Cooker-Monster** [comment]
I can't figure out how tricky this point is inside the existing code (I'm mainly an enthusiastic user of the framework) but to sumup my way of thinking, every time I want to promote Loopback to dev teams, I'm facing a little desapointment when POST returns 200 in my demo. So debate is always fructeous, but I would like to find a quick workaround to fix this important point.

**Comment by bajtos** [comment]
Few more ideas to consider:

When returning `201 Created`, the response should include a `Location` header providing a URL of the newly created entity to clients. See https://www.w3.org/Protocols/rfc2616/rfc2616-sec10.html#sec10.2.2, emphasis is mine:

> The newly created resource can be referenced by the URI(s) returned in the entity of the response, **with the most specific URI for the resource given by a Location header field.**

If we had support for configurable response types/formats already implemented (see https://github.com/strongloop/loopback-next/issues/436), then the problem with 201 status code  could be solved by changing the template of `create` controller methods to something like below:

```ts
  @post('/products')
  async create(@requestBody() obj: Product)
    : Promise<ControllerResponse<Product>> {
    const model = await this.productRepository.create(obj);
    return {
      statusCode: 201,
      // FIXME: Do not hard-code `model.id`, find the actual PK property name.
      // FIXME: Can Location contain a path only, with no hostname and scheme?
      headers: {location: `/products/${model.id}`},
      body: model,
    }
  }
```

**Comment by jannyHou** [comment]
I created a PoC PR for approach "reading success statusCode from `response` spec": https://github.com/strongloop/loopback-next/pull/1260, summarized our discussion in it. Maybe we can continue discussion in that PR then post the solution here.

**Comment by bajtos** [comment]
FWIW, the latest version of HTTP/1.1 spec allows the `Location` header to contain a relative path.

See https://tools.ietf.org/html/rfc7231#section-7.1.2, emphasis is mine:

> The "Location" header field is used in some responses to refer to a
   specific resource in relation to the response.  The type of
   relationship is defined by the combination of request method and
   status code semantics.
>
>     Location = URI-reference
>
>   The field value consists of a single URI-reference.  **When it has the
   form of a relative reference ([RFC3986], Section 4.2), the final
   value is computed by resolving it against the effective request URI
   ([RFC3986], Section 5).**

See also https://en.wikipedia.org/wiki/HTTP_location.

**Comment by dhmlau** [comment]
Seems to be a duplicate of #436.

**Comment by bajtos** [comment]
Definitely not a duplicate. We may need #436 to be implemented first before we can return 201 with a Location header.

I am removing this story from GA and moving it to _[EPIC] REST layer improvements (post-GA) #1452_

**Comment by lajosf** [comment]
Dear @bajtos , as I can see the "stale bot" gonna close this thread, because there is no activity. As I can see it is still a problem, not solved yet and may effect more developers.

**Comment by bajtos** [comment]
@lajosf I am afraid we don't have enough bandwidth to implement this feature soon. Would you like to contribute it yourself?

**Comment by lajosf** [comment]
Dear Bajtos,
Thank you for the quick answer! After I wrote the above comment I made workaround and works fine.
I have no option to contribute, but thank you (one half of my life is reserved by the company, the other is by my family :))!
